dialogue-target/src/main/java/com/palantir/dialogue/DeserializerArgs.java

```diff
@@ -0,0 +1,98 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue;
+
+import com.google.common.collect.ImmutableMap;
+import com.palantir.logsafe.Preconditions;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public final class DeserializerArgs<T> {
+    private final TypeMarker<T> baseType;
+    private final TypeMarker<? extends T> successType;
+    private final ImmutableMap<String, TypeMarker<? extends T>> errorNameToTypeMarker;
+
+    private DeserializerArgs(
+            TypeMarker<T> baseType,
+            TypeMarker<? extends T> successType,
+            ImmutableMap<String, TypeMarker<? extends T>> map) {
+        this.baseType = baseType;
+        this.successType = successType;
+        this.errorNameToTypeMarker = map;
+    }
+
+    public static <T> Builder<T> builder() {
+        return new Builder<>();
+    }
+
+    public static final class Builder<T> {
+        private boolean buildInvoked = false;
+        private @Nullable TypeMarker<T> baseType;
+        private @Nullable TypeMarker<? extends T> successType;
+        private final Map<String, TypeMarker<? extends T>> errorNameToTypeMarker;
+
+        private Builder() {
+            this.errorNameToTypeMarker = new HashMap<>();
+        }
+
+        public Builder<T> baseType(@Nonnull TypeMarker<T> baseT) {
+            checkNotBuilt();
+            this.baseType = Preconditions.checkNotNull(baseT, "base type must be non-null");
+            return this;
+        }
+
+        public Builder<T> success(@Nonnull TypeMarker<? extends T> successT) {
+            checkNotBuilt();
+            this.successType = Preconditions.checkNotNull(successT, "success type must be non-null");
+            return this;
+        }
+
+        public Builder<T> error(@Nonnull String errorName, @Nonnull TypeMarker<? extends T> errorT) {
+            checkNotBuilt();
+            this.errorNameToTypeMarker.put(
+                    Preconditions.checkNotNull(errorName, "error name must be non-null"),
+                    Preconditions.checkNotNull(errorT, "error type must be non-null"));
+            return this;
+        }
+
+        public DeserializerArgs<T> build() {
+            checkNotBuilt();
+            Preconditions.checkNotNull(baseType, "base type must be set");
+            Preconditions.checkNotNull(successType, "success type must be set");
```

Comment:
fysa @carterkozak, I'm thinking through how we can handle endpoints that don't return anything but can return errors. This `success` type doesn't always need to be set.

Reply:
Yeah. I was initially thinking that if it's an empty type, it wouldn't need to be provided by the caller (via `DeserializerArgs`). That doesn't work though: the caller needs to get _some_ object to indicate a successful result.